downloads/release-notes/1.2.7/RELEASE_NOTE.md

```diff
@@ -22,9 +22,92 @@ However, you need to use the following commands instead of the above if you have
 
 If you have not installed jBallerina, then download the [installers](https://ballerina.io/downloads/) to install.
 
+## Language
+
 ## Standard Library
-- 
 
-## Dev Tools
-- 
+### I/O
+
+When a CSV file has an empty record, the previous implementation throws a null pointer exception and it was fixed with this release.
+
+## Developer Tools
+
+### Test Framework
+
+#### Object Mocking
+
+Object mocking enables controlling the values of member variables and the behavior of the member functions of an object. The following improvmenets are introduced with regard to this.
+
+- Ability to create a test double, which provides an equivalent mock in place of the real object
+- Capability of stubbing the member function or member variable
+
+Object mocking is done by using the following functions.
+
+- The `test:mock()` and `test:prepare()` functions are used to initialize the mocking capability.
+- The `test:prepare()` function allows to use the associated mocking functions like `thenReturn()`, `thenReturnSequence()`, `doNothing()`, and `withArguments()`.
+
+#### Function Mocking
+
+- The `MockFunction` object is added to handle function mocking. `MockFunction` objects are defined by attaching the `@test:Mock` to the `MockFunction` object to specify the function to mock.
+- The mocking API also supports scoping and stubbing of mock functions that are declared for functions in imported modules.
+
+Function mocking is done by using the following functions.
+
+- The `test:when(mockObj)` function is used to initialize the mocking capability within a particular test case.
+- This allows you to use the associated mocking functions like `call()`, `thenReturn()`, and `withArguments()`.
+
+### Improvements related to code actions
+
+#### Introducing an AI-based `Data Mapping` Code Action
```

Comment:
This should be release with SwanLake not 1.2.7 - we need to remove this from here

Reply:
Done.